### PR TITLE
Create an Options page for the Cloud Explorer.

### LINKED ISSUE #562: Add option page for Pub/Sub filtered topics.
Add an option page to Google Cloud Tools options in Visual Studio options so users can update the list of filtered Pub/Sub topics themselves.

### PR DESCRIPTION
**Pull request by ILMTitan**
Add the ability to select which topics are filtered from the cloud explorer.

Fixes #562

### CODE REVIEW

File: GoogleCloudExtension/GoogleCloudExtension/Analytics/Events/NewInstallEvent.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace GoogleCloudExtension.Analytics.Events
     /// </summary>
     internal static class NewInstallEvent
     {
-        private const string NewInstallEventName = "newInstall";
+        internal const string NewInstallEventName = "newInstall";
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Analytics/Events/NewInstallEvent.cs]: This change appears unrelated to the fix.

File: GoogleCloudExtension/GoogleCloudExtension/Analytics/Events/UpgradeEvent.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ namespace GoogleCloudExtension.Analytics.Events
     /// </summary>
     internal static class UpgradeEvent
     {
-        private const string UpgradeEventName = "upgrade";
+        internal const string UpgradeEventName = "upgrade";
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Analytics/Events/UpgradeEvent.cs]: Same here, unrelated to the fix.

File: GoogleCloudExtension/GoogleCloudExtension/Analytics/EventsReporterWrapper.cs [hunk 3]
```diff
@@ -30,14 +30,18 @@ internal static class EventsReporterWrapper
...
+        private static IEventsReporter Reporter => ReporterLazy.Value;
+
 
         /// <summary>
         /// Used by unit test to prevent analytics from running.
         /// </summary>
         public static void DisableReporting()
         {
-            s_reporter = new Lazy<IEventsReporter>(() => null);
+            ReporterLazy = new Lazy<IEventsReporter>(() => null);
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Analytics/EventsReporterWrapper.cs]: These changes are unrelated to the new options page, please split into their own PR for ease of review.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            
+            <DataGrid.Columns>
+                <DataGridTextColumn
+                    Header="{x:Static ext:Resources.CloudExplorerOptionsPubSubTopicsFilterColumHeader}"
+                    IsReadOnly="False"
+                    Binding="{Binding Value, UpdateSourceTrigger=PropertyChanged}"
+                    Width="*"/>
+            </DataGrid.Columns>
+        </DataGrid>
+        <Button
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml]: Can we use the standard styles for the `Button` so it looks like other buttons in the extension?

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        /// Creates and initializes a new <see cref="CloudExplorerOptionsPage"/>.
+        /// </summary>
+        /// <param name="parentModel"></param>
+        public CloudExplorerOptionsPage(ICloudExplorerOptions parentModel)
+        {
+            ViewModel = new CloudExplorerOptionsPageViewModel(parentModel.ResetSettings);
+            DataContext = ViewModel;
+            parentModel.SavingSettings += UpdateBindingSources;
+            AddHandler(UIElementDialogPage.DialogKeyPendingEvent, new RoutedEventHandler(OnDialogKeyPending));
+            InitializeComponent();
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs]: `InitializeComponent()` should be the first thing.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        /// <summary>
+        /// Handles the <see cref="UIElementDialogPage.DialogKeyPendingEvent"/>,
+        /// allowing this WPF element to catch enter and escape keys during editing mode.
+        /// </summary>
+        /// <param name="sender">Unused.</param>
+        /// <param name="args">
+        /// Updates the <see cref="RoutedEventArgs.Handled"/> property.
+        /// </param>
+        private void OnDialogKeyPending(object sender, RoutedEventArgs args)
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs]: Shouldn't `enter` and `esc` be handled by the `DataGrid` or any other input controls to allow the user to commit, and cancel out, of the editing operations?

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+using GoogleCloudExtension.Utils;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace GoogleCloudExtension.CloudExplorer.Options
+{
+    public class CloudExplorerOptionsPageViewModel : ViewModelBase
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: No class comments or public method comments.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        private ObservableCollection<EditableModel<string>> _pubSubTopicFilters;
+
+        public IEnumerable<EditableModel<string>> PubSubTopicFilters
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: Why not just make the property of type `ObservableCollection`? You can manipulate that class and the changes should be reflected in the UI. No need to create a new fresh copy of the collection every time you set it.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/ICloudExplorerOptions.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace GoogleCloudExtension.CloudExplorer.Options
+{
+    /// <summary>
+    /// The options that affect the Cloud Explorer window.
+    /// </summary>
+    public interface ICloudExplorerOptions
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/ICloudExplorerOptions.cs]: I guess that this interface will contain more settings should we add them?

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+// limitations under the License.
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+
+namespace GoogleCloudExtension.CloudExplorer.Options
+{
+    public class SerializableCloudExplorerOptions
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs]: No class comments.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        /// A json string serialized version of the PubSubTopicFilters.
+        /// Used for saving and loading to the settings storage.
+        /// </summary>
+        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
+        // ReSharper disable once UnusedMember.Global
+        public string PubSubTopicFiltersJsonString
+        {
+            get
+            {
+                return _parentOptions.PubSubTopicFilters == null ?
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs]: I typically use `JsonConverter` to do the transition between json <-> object when I can use all of the defaults. I find it more readable.

File: GoogleCloudExtension/GoogleCloudExtension/Resources.resx [hunk 12]
```diff
@@ -2942,4 +2942,16 @@
   <data name="WizardTemplateChooserTargetAspFrameworkLabel" xml:space="preserve">
     <value>Target _ASP.NET Framework:</value>
   </data>
+  <data name="CloudExplorerOptionsPubSubTopicsLabel" xml:space="preserve">
+    <value>Pub/Sub Topics to hide:</value>
+  </data>
+  <data name="CloudExplorerOptionsPubSubTopicsLabelHelp" xml:space="preserve">
+    <value>If any of the patterns match a topic's full name (e.g. projects/project-id/topics/topic-name), that topic will not be shown in the Cloud Explorer.</value>
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Resources.resx]: This reads like a double negative of sorts. How about something like _Regular expressions for topics to exclude from the cloud explorer_ or something like that?

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+{
+
+    /// <summary>
+    /// An <see cref="IEditableObject"/> for a single object.
+    /// </summary>
+    /// <typeparam name="T">The type this EditableModel contains.</typeparam>
+    public class EditableModel<T> : Model, IEditableObject
+    {
+        private T _value;
+        private T _uneditedRegex;
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: The field `_uneditedRegex` name is problably and hold over from when this class was more specific. Should use a more generic name now.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        /// <param name="input">The object to wrap.</param>
+        /// <returns>A new <see cref="EditableModel{T}"/> wrapping the input.</returns>
+        public static EditableModel<T> Of<T>(T input) => new EditableModel<T>(input);
+
+        /// <summary>
+        /// Wraps the elements of an <see cref="IEnumerable{T}"/> in <see cref="EditableModel{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">The type to wrap.</typeparam>
+        /// <param name="input">The objects to wrap.</param>
+        public static IEnumerable<EditableModel<T>> ToEditableModels<T>(this IEnumerable<T> input) =>
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: I think that in this method you want to be eager and return a list with the models all created. Otherwise you risk recreating them unexpectedly.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+using System.Linq;
+
+namespace GoogleCloudExtension.Utils
+{
+
+    /// <summary>
+    /// An <see cref="IEditableObject"/> for a single object.
+    /// </summary>
+    /// <typeparam name="T">The type this EditableModel contains.</typeparam>
+    public class EditableModel<T> : Model, IEditableObject
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: While I appreciate the generic class, this object is of very limited use. Only useable when the model is a single value such as a string. The `IEditableObject` interface can also be used for general models, with more than one property. For that there's really no generic way to write a class, generating code would be better.

I say this because you might find that we will not be using this generic much.

File: GoogleCloudExtension/GoogleCloudExtension/packages.config [hunk 16]
```diff
@@ -16,6 +16,7 @@
   <package id="Google.Apis.ServiceManagement.v1" version="1.29.1.981" targetFramework="net452" />
   <package id="Google.Apis.SQLAdmin.v1beta4" version="1.29.1.949" targetFramework="net452" />
   <package id="Google.Apis.Storage.v1" version="1.29.1.988" targetFramework="net452" />
+  <package id="JetBrains.Annotations" version="11.1.0" targetFramework="net452" />
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/packages.config]: What is this new package for? What is the license?

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs (hunk 5 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs]: Why? It doesn't seem to matter.

Fixed.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs (hunk 6 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs]: The problem is the hosting non-WPF `UIElementDialogPage` captures non-alphanumeric keyboard input for its own purposes (e.g. closing or canceling the the whole options dialog). The routed event `UIElementDialogPage.DialogKeyPendingEvent` allows the hosted WPF control to override that behavior, by setting the event to `Handled = true`. With that done, the standard WPF events are triggered normally (which the `DataGrid` does handle).

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs (hunk 7 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: Fixed.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs (hunk 8 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/ICloudExplorerOptions.cs (hunk 9 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/ICloudExplorerOptions.cs]: Yes, it would.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs (hunk 10 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs (hunk 11 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/SerializableCloudExplorerOptions.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/Resources.resx (hunk 12 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/Resources.resx]: Changed to 

*Topics with a full name (e.g. projects/project-id/topics/topic-name) that matches one of these regex patters will be hidden in the Cloud Explorer.*

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs (hunk 13 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: Fixed.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs (hunk 14 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: Only if a caller iterates the enumerable twice, which is an error on their part.

File: GoogleCloudExtension/GoogleCloudExtension/packages.config (hunk 16 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/packages.config]: It adds some attributes that assist with Resharper tooling. The `[UsedImplicitly]` attribute on the `EditableModel()` constructor comes from this package, and prevents Resharper from complaining that the constructor not used.

The license on the package is the [MIT license](https://raw.githubusercontent.com/JetBrains/ExternalAnnotations/master/LICENSE.md).
See the package on [NuGet](https://www.nuget.org/packages/JetBrains.Annotations).


Looking at it again, I don't need it. Removing the package.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs (hunk 14 above)
Comment (ivannaranjo) [reply GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: It is never an error to enumerate enumerables, that's what they are for.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs (hunk 15 above)
Comment (ivannaranjo) [reply GoogleCloudExtension/GoogleCloudExtension/Utils/EditableModel.cs]: Yep, but we can undo it or de emphasize it.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs (hunk 6 above)
Comment (ivannaranjo) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml.cs]: I see.

**Comment by ivannaranjo** [comment]
I think that we need to rethink this and possibly add this as an option for the cloud explorer itself and have it by project, etc... I do not think that we can have a global list of exclusions that can apply to every single project the user works with. In short, I think that the global options are not the right place to put this options.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml (hunk 4 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPage.xaml]: Because I am importing *DefaultStylesResource.xaml*, rather than *ThemingResources.xaml*, this `Button` by default has the standard style defined in *TemingResources.xaml*.

**Comment by ILMTitan** [comment]
I added some context menus to make it easier to access the options from the cloud explorer. I also created Issue #864 to track making a per-GCP Project version of these options.

File: GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs [hunk 17]
```diff
@@ -154,7 +155,17 @@ public class PubsubDataSource : DataSourceBase<PubsubService>, IPubsubDataSource
         /// </summary>
         public static string GetPathLeaf(string path)
         {
-            return path.Substring(1 + path.LastIndexOf("/", StringComparison.Ordinal));
+            return GetPathSections(path).LastOrDefault();
+        }
+
+        public static string GetTopicProject(string path)
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs]: Nit, method comment.

Comment (ivannaranjo) [reply GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs]: Specially important to add examples of the transformation that this method produces.

File: GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs [hunk 18]
```diff
@@ -19,6 +19,7 @@
 using GoogleCloudExtension.DataSources;
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading.Tasks;
 
 namespace GoogleCloudExtension.CloudExplorerSources.PubSub
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs]: Consider at some point having a full resource parser. The names used by the API follow a very particular format that is only _path like_ enough to allow these manipulations to succeed. This could be a future enhancement.

File: GoogleCloudExtension/GoogleCloudExtension/Analytics/EventsReporterWrapper.cs [hunk 19]
```diff
@@ -75,16 +75,16 @@ public static void AnalyticsOptInStateChanged()
         /// Called to report an interesting event to analytics. If there's a queue of events it will be
         /// flushed as well.
         /// </summary>
-        public static Action<AnalyticsEvent> ReportEvent { get; internal set; } = (AnalyticsEvent eventData) =>
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/Analytics/EventsReporterWrapper.cs]: This change doesn't seem that should be part of the change for the cloud explorer options page.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptions.cs [hunk 20]
```diff
@@ -45,7 +51,11 @@ public class CloudExplorerOptions : UIElementDialogPage, ICloudExplorerOptions
         public IEnumerable<string> PubSubTopicFilters
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptions.cs]: Since you are using an `ObservableCollection<T>` you should make the property that type. And then add/remove from the collection. Creating a new instance when you set the value breaks the observable model and makes the binding expression more expensive for no gain.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs [hunk 21]
```diff
@@ -14,26 +14,21 @@
...
 {
+    /// <summary>
+    /// View model for the <see cref="CloudExplorerOptionsPage"/>.
+    /// </summary>
     public class CloudExplorerOptionsPageViewModel : ViewModelBase
     {
         private ObservableCollection<EditableModel<string>> _pubSubTopicFilters;
 
-        public IEnumerable<EditableModel<string>> PubSubTopicFilters
+        public ObservableCollection<EditableModel<string>> PubSubTopicFilters
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: A property of type `ObservableCollection<T>` should typically be a readonly property, since the change notifications are coming from manipulating the collection. Therefore you should just assign the value to the property during the construction of the object.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs [hunk 22]
```diff
@@ -14,26 +14,21 @@
...
         {
             get { return _pubSubTopicFilters; }
-            set
-            {
-                SetValueAndRaise(
-                    ref _pubSubTopicFilters,
-                    new ObservableCollection<EditableModel<string>>(
-                        value ?? Enumerable.Empty<EditableModel<string>>()));
-            }
+            set { SetValueAndRaise(ref _pubSubTopicFilters, value); }
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: Since you are using the `ObservableCollection<T>` this call to `SetValueAndRaise` is really unnecessary.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs [hunk 23]
```diff
@@ -148,6 +158,19 @@ public override void InvalidateProjectOrAccount()
             _dataSource = new Lazy<IPubsubDataSource>(_dataSourceFactory);
         }
 
+        /// <summary>
+        /// Opens an external browser to the given url.
+        /// </summary>
+        /// <param name="url"></param>
+        public void OpenBrowser(string url)
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: If you are just opening a browser on a URL then you can just use `Process.Start(url)`, which will do the same thing and much simpler, including the opening the default browser registered by the user. We use this all over the extension.

File: GoogleCloudExtension/GoogleCloudExtension/GoogleCloudExtensionPackage.cs [hunk 24]
```diff
@@ -257,9 +259,17 @@ protected override void Initialize()
 
         #region User Settings
 
-        public AnalyticsOptionsPage AnalyticsSettings => GetDialogPage<AnalyticsOptionsPage>();
+        public AnalyticsOptions AnalyticsSettings => GetDialogPage<AnalyticsOptions>();
 
-        public T GetDialogPage<T>() where T : DialogPage => (T) GetDialogPage(typeof(T));
+        public T GetDialogPage<T>() where T : DialogPage
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/GoogleCloudExtensionPackage.cs]: Method comments.

File: GoogleCloudExtension/GoogleCloudExtension/Analytics/EventsReporterWrapper.cs (hunk 19 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/Analytics/EventsReporterWrapper.cs]: I'm not sure where it is getting the code on the left here. The code on the right is what exists in both branches, master issue#562.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptions.cs (hunk 20 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptions.cs]: I am not sure what you think is being broken.
I am also not sure why you think clearing and then adding back to the `ObservableCollection` is less expensive. Just setting the property to a new collection will trigger a single event (O(1)), which triggers the UI to update (a O(n) operation). Mutating the collection, however, triggers many events. First, an event when the collection is cleared. Then because it doesn't have an `AddRange`, each item would have to be added individually. That means a new event on each add (O(n)). Each of these events could trigger a UI update (O(n)), for a total cost of O(n<sup>2</sup>).

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs (hunk 21 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorer/Options/CloudExplorerOptionsPageViewModel.cs]: While it could be a readonly property, I'm not sure where you are getting that it should be, or even typically is. Doing so would avoid the need for a property changed notification mechanism, but given the [Docs for `ObservableCollection`](https://msdn.microsoft.com/en-us/library/ms668604(v=vs.110).aspx#Anchor_8) has a note about using the `INotifyPropertyChanged` interface, it seems pretty standard to use that mechanism as well.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs (hunk 23 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Doing this allows me to mock and register the `IVsWebBrowsingService` in unit tests. It allows me to avoid having an ugly `internal Action<string> StartProcess = Process.Start` variable, and is more idiomatic to Visual Studio.

File: GoogleCloudExtension/GoogleCloudExtension/GoogleCloudExtensionPackage.cs (hunk 24 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/GoogleCloudExtensionPackage.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs (hunk 17 above)
Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension.DataSources/PubSubDataSource.cs]: Done.

**Comment by ILMTitan** [comment]
PTAL.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs (hunk 23 above)
Comment (ivannaranjo) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Please don't, use the same way of opening browsers as the rest of the code. If you want to mock the way the browser is opened do it the existing way, it is not ugly, it is a mock.

Comment (ILMTitan) [reply GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Changed.

**Comment by ILMTitan** [comment]
Done.
PTAL.
